src/ibmca_ec.c

```diff
@@ -0,0 +1,918 @@
+        goto end;
+    }
+
+    /* Allocate memory for shared secret z, will be freed by caller */
+    if ((z_buf = OPENSSL_malloc(privlen)) == NULL) {
+        IBMCAerr(IBMCA_F_IBMCA_ECDH_COMPUTE_KEY, IBMCA_R_EC_INTERNAL_ERROR);
+        goto end;
+    }
+
+    /* Calculate shared secret z */
+    rc = p_ica_ecdh_derive_secret(ibmca_handle, ica_privkey, ica_pubkey, z_buf,
+                                  privlen);
+    if (rc != 0) {
+        /* Possibly no suitable adapter. */
+        OPENSSL_free(z_buf);
+
+ #ifdef OLDER_OPENSSL
+        goto end;
+ #else
+        EC_KEY_METHOD_get_compute_key(ossl_ec, &compute_key_sw);
+        if (compute_key_sw == NULL) {
+            IBMCAerr(IBMCA_F_ICA_ECDH_DERIVE_SECRET, rc);
+            goto end;
+        }
+
+        ret = compute_key_sw(pout, poutlen, pub_key, ecdh);
+        goto end;
+ #endif
+    }
+
+    *pout = z_buf;
+    *poutlen = privlen;
+
+    ret = 1;
+
+end:
+    p_ica_ec_key_free(ica_pubkey);
+    p_ica_ec_key_free(ica_privkey);
+
+    return ret;
+}
+
+ #ifndef OLDER_OPENSSL
+/**
+ * ECDSA signing method (replaces ossl_ecdsa_sign).
+ *
+ * returns 1 if success
+ *         0 if error
+ */
+int ibmca_ecdsa_sign(int type, const unsigned char *dgst, int dlen,
+             unsigned char *sig_array, unsigned int *siglen,
+             const BIGNUM *kinv, const BIGNUM *r, EC_KEY *eckey)
+{
+    ECDSA_SIG *sig;
+    const BIGNUM *bn_r, *bn_s;
+    const EC_GROUP *group;
+    int n, i, r_len;
+
+    /* Check parms: precomputed (k,r) are not supported by ibmca */
+    if (kinv != NULL || r != NULL) {
+        IBMCAerr(IBMCA_F_IBMCA_ECDSA_SIGN, IBMCA_R_EC_INVALID_PARM);
+        return 0;
+    }
+
+    /* Create signature */
+    sig = ibmca_ecdsa_sign_sig(dgst, dlen, NULL, NULL, eckey);
+    if (sig == NULL) {
+        IBMCAerr(IBMCA_F_IBMCA_ECDSA_SIGN, IBMCA_R_EC_INTERNAL_ERROR);
+        *siglen = 0;
+        return 0;
+    }
+
+    /* Determine r-length */
+    if ((group = EC_KEY_get0_group(eckey)) == NULL) {
+        IBMCAerr(IBMCA_F_IBMCA_ECDSA_SIGN, IBMCA_R_EC_INTERNAL_ERROR);
+        return 0;
+    }
+    r_len = (EC_GROUP_get_degree(group) + 7) / 8;
+
+    /* Get (r,s) from ECDSA_SIG */
+    ECDSA_SIG_get0(sig, &bn_r, &bn_s);
+
+    /* Format r as byte array with leading 0x00's if necessary */
+    n = r_len - BN_num_bytes(bn_r);
+    for (i = 0; i < n; i++)
+        sig_array[i] = 0x00;
+    BN_bn2bin(bn_r, &(sig_array[n]));
+
+    /* Format s as byte array with leading 0x00's if necessary */
+    n = r_len - BN_num_bytes(bn_s);
+    for (i = 0; i < n; i++)
+        sig_array[r_len + i] = 0x00;
+    BN_bn2bin(bn_s, &(sig_array[r_len+n]));
+
+    /* Create DER encoding */
+    *siglen = i2d_ECDSA_SIG(sig, &sig_array);
+
+    return 1;
+}
+ #endif
+
+/**
+ * ECDSA signing method (replaces ossl_ecdsa_sign_sig).
+ *
+ * @return pointer to a ECDSA_SIG structure or NULL if an error occurred
+ */
+ECDSA_SIG *ibmca_ecdsa_sign_sig(const unsigned char *dgst, int dgst_len,
```

Comment:
This and the previous function have the exact same leading comment but doing different things.